keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1330 @@
+   possible to disrupt the pods during a high load without experiencing application downtime. If
+   the minimum number of pods is 1, PDBs cannot be used without blocking the node drain. This has
+   been discussed in issue [kubernetes/kubernetes#93476](https://github.com/kubernetes/kubernetes/issues/93476)
+3. Graceful deletion of DaemonSet pods is currently only supported as part of (Linux) graceful node
+   shutdown. The length of the shutdown is again not application specific and is set cluster-wide
+   (optionally by priority) by the cluster admin. This does not take into account 
+   `.spec.terminationGracePeriodSeconds` of each pod and may cause premature termination of
+   the application. This has been discussed in issue [kubernetes/kubernetes#75482](https://github.com/kubernetes/kubernetes/issues/75482)
+   and in issue [kubernetes-sigs/cluster-api#6158](https://github.com/kubernetes-sigs/cluster-api/issues/6158).
+4. There are cases during a node shutdown, when data corruption can occur due to premature node
+   shutdown. It would be great if applications could perform data migration and synchronization of 
+   cached writes to the underlying storage before the pod deletion occurs. This is not easy to
+   quantify even with pod's `.spec.shutdownGracePeriod`, as the time depends on the size of the data
+   and the speed of the storage. This has been discussed in issue [kubernetes/kubernetes#116618](https://github.com/kubernetes/kubernetes/issues/116618)
+   and in issue [kubernetes/kubernetes#115148](https://github.com/kubernetes/kubernetes/issues/115148).
+5. There is not enough metadata about why the node drain was requested. This has been discussed in
+   issue [kubernetes/kubernetes#30586](https://github.com/kubernetes/kubernetes/issues/30586).
+
+Approaches and workarounds used by other projects to deal with these shortcomings:
+- https://github.com/medik8s/node-maintenance-operator uses a declarative approach that tries to
+  mimic `kubectl drain` (and uses kubectl implementation under the hood).
+- https://github.com/kubereboot/kured performs automatic node reboots and relies on `kubectl drain`
+  implementation to achieve that.
+- https://github.com/strimzi/drain-cleaner prevents Kafka or ZooKeeper pods from being drained
+  until they are fully synchronized. Implemented by intercepting eviction requests with a
+  validating admission webhook. The synchronization is also protected by a PDB with the
+  `.spec.maxUnavailable` field set to 0. See the experience reports for more information.
+- https://github.com/kubevirt/kubevirt intercepts eviction requests with a validating admission
+  webhook to block eviction and to start a virtual machine live migration from one node to another.
+  Normally, the workload is also guarded by a PDB with the `.spec.minAvailable` field set to 1.
+  During the migration the value is increased to 2.
+
+Experience Reports:
```

Comment:
IMHO a lot of what is in this section is pretty anecdotal. It is from people/projects that were vocal but its unclear if what what they are describing is common.

OTOH what is extremely commonly used an notably absent here is how the three big CSPs (AWS, Azure, GCP) do node draining in their respective nodepool upgrade functionality of their managed k8s offerings.

Reply:
I have added the CSP use cases. However most of them use the Cluster Autoscaler.

Experience Reports are only part of the puzzle. There are a lot of relevant upstream issues and customer issue that experience these common problems. And blogs written by the users help to point that out.

@alvaroaleman please let me know if you have other sources you would like to add?